cmd/admin-handlers.go

```
@@ -1360,6 +1361,84 @@ func (a adminAPIHandlers) ServerHardwareInfoHandler(w http.ResponseWriter, r *ht
 	}
 }
 
+// OBDInfoHandler - GET /minio/admin/v2/obdinfo
+// ----------
+// Get server on-board diagnostics
+func (a adminAPIHandlers) OBDInfoHandler(w http.ResponseWriter, r *http.Request) {
+	ctx := newContext(r, w, "OBDInfo")
+	objectAPI, _ := validateAdminReq(ctx, w, r, iampolicy.OBDInfoAdminAction)
+	if objectAPI == nil {
+		return
+	}
+
+	if globalIsDistXL {
+		zones, ok := objectAPI.(*xlZones)
+		if !ok {
+			panic("Unreachable")
+		}
+
+		// Hold a lock so only one server performs on-board diagnostics
+		obdLock := zones.NewNSLock(ctx, minioMetaBucket, "obd")
+		if err := obdLock.GetLock(newDynamicTimeout(5*time.Second, 2*time.Second)); err != nil {
+			writeErrorResponseJSON(ctx, w, errorCodes.ToAPIErr(ErrOBDLockTimeout), r.URL)
+			return
+		}
+		defer obdLock.Unlock()
+	} else {
+		fsObjects, ok := objectAPI.(*FSObjects)
+		if !ok {
+			panic("Unreachable")
+		}
+
+		// Hold a lock so only one server performs on-board diagnostics
+		obdLock := fsObjects.NewNSLock(ctx, minioMetaBucket, "obd")
+		if err := obdLock.GetLock(newDynamicTimeout(5*time.Second, 2*time.Second)); err != nil {
+			writeErrorResponseJSON(ctx, w, errorCodes.ToAPIErr(ErrOBDLockTimeout), r.URL)
+			return
+		}
+		defer obdLock.Unlock()
+
+	}
```

@wlan0, the handler shouldn't have logic specific to fs or xl. Suggest defining a GetOBDInfo function in object-api-interface.go and move the functionality here to xl/fs layers. For gateway, you will just need to  raise a  NotImplemented error in gateway-unsupported.go

---

I'll do that

---

done